### PR TITLE
Evaporator Refinement

### LINKED ISSUE #551: Separate condenser from evaporator
Currently the evaporator model for the MVC model incorporates a feed side evaporator and a condenser. This causes issues for initialization and a better approach is to have the evaporator only contain the feed side. 

Separating the condenser from the evaporator will require a method to connect the condenser to the evaporator in the flowsheet.

### PR DESCRIPTION
**Pull request by carson-tucker**
## Fixes/Resolves:
Fixes #551

## Summary/Motivation:
The MVC evaporator will be easier to initialize if it is separate from the condenser. 

## Changes proposed in this PR:
- Separates condenser from the evaporator
- Adds method to connect condenser to evaporator in flowsheet
- Updates tests

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by TimBartholomew** [comment]
@bknueven @lbianchi-lbl - Right now this PR just adds one blank line, I was helping Carson create her first PR. I'm surprised that all the tests are failing. Do you guys have any insight? 

Edit: Oops just needed to rerun the tests again. We know why black is failing.

**Comment by lbianchi-lbl** [comment]
> @bknueven @lbianchi-lbl - Right now this PR just adds one blank line, I was helping Carson create her first PR. I'm surprised that all the tests are failing. Do you guys have any insight?
> 
> Edit: Oops just needed to rerun the tests again. We know why black is failing.

@TimBartholomew I think the error might have been caused by a short-lived bug in one of our indirect dependencies: https://github.com/ipython/traitlets/issues/741. This has since been fixed, so in case it showed up in other PRs, re-running the failing jobs should be enough.

File: watertap/unit_models/mvc/components/complete_condenser.py [hunk 1]
```diff
@@ -253,6 +255,19 @@ def initialize(
...
+        if degrees_of_freedom(blk) != 0:
+            if heat != None:
+                blk.control_volume.heat.fix(heat)
+                under_constrained_flag = True
+            else:
+                raise RuntimeError(
+                    "The model has {} degrees of freedom rather than 0 for initialization."
+                    " This error suggests that an outlet condition has not been fixed"
+                    " for initialization.".format(degrees_of_freedom(blk))
+                )
```
Comment (bknueven) [inline watertap/unit_models/mvc/components/complete_condenser.py]: This control flow doesn't make much sense to me.

I think `under_constrained_flag` should be renamed `fixed_heat`, and more critically, if the degrees of freedom is greater than 1, and the caller has provided a value for `heat`, we will continue initializing the model.

It seems we need to raise and error if the degrees of freedom is greater than 1, and if it's exactly 1 then we can do this check for heat.

File: watertap/unit_models/mvc/components/evaporator.py [hunk 2]
```diff
@@ -430,70 +430,53 @@ def initialize(
...
+            if delta_temperature_in != None and delta_temperature_out != None:
+                blk.delta_temperature_in.fix(delta_temperature_in)
+                blk.delta_temperature_out.fix(delta_temperature_out)
+                under_constrained_flag = True
+            else:
+                raise RuntimeError(
+                    "The model has {} degrees of freedom rather than 0 for initialization."
+                    " This error suggests that temperature differences have not been fixed"
+                    " for initialization.".format(degrees_of_freedom(blk))
+                )
```
Comment (bknueven) [inline watertap/unit_models/mvc/components/evaporator.py]: There's a similar issue here.

**Review by bknueven — CHANGES_REQUESTED** [review]
I haven't gotten a chance to do a thorough review, but I noticed a problem in the initialization logic.

File: watertap/unit_models/mvc/tests/test_mvc.py [hunk 3]
```diff
@@ -52,14 +54,16 @@ def build(m):
     # Compressor
     m.fs.compressor = Compressor(default={"property_package": m.fs.properties_vapor})
```
Comment (adam-a-a) [inline watertap/unit_models/mvc/tests/test_mvc.py:55]: Out of curiosity, what are the main differences between this compressor and the IDAES compressor model? Why not use the IDAES version?

File: watertap/unit_models/mvc/tests/test_mvc.py [hunk 4]
```diff
@@ -52,14 +54,16 @@ def build(m):
...
 
+    # Condenser
+    m.fs.condenser = Condenser(default={"property_package": m.fs.properties_vapor})
```
Comment (adam-a-a) [inline watertap/unit_models/mvc/tests/test_mvc.py:58]: Similarly, does IDAES not have a Condenser unit model available? Just want to refresh my memory on why it was necessary to make custom models for these.

File: watertap/unit_models/mvc/tests/test_mvc.py [hunk 5]
```diff
@@ -52,14 +54,16 @@ def build(m):
...
     # Connections
     m.fs.s01 = Arc(
         source=m.fs.evaporator.outlet_vapor, destination=m.fs.compressor.inlet
     )
-    m.fs.s02 = Arc(
-        source=m.fs.compressor.outlet, destination=m.fs.evaporator.inlet_condenser
-    )
+    m.fs.s02 = Arc(source=m.fs.compressor.outlet, destination=m.fs.condenser.inlet)
     TransformationFactory("network.expand_arcs").apply_to(m)
+    m.fs.evaporator.connect_to_condenser(m.fs.condenser)
```
Comment (adam-a-a) [inline watertap/unit_models/mvc/tests/test_mvc.py:66]: Additionally, why not make an MVC unit model file that encapsulates the condenser, compressor, and evaporator together?

File: watertap/unit_models/mvc/components/tests/test_complete_condenser.py [hunk 6]
```diff
@@ -50,22 +51,17 @@ def test_complete_condense():
     assert_units_consistent(m)
     assert degrees_of_freedom(m) == 0
```
Comment (adam-a-a) [inline watertap/unit_models/mvc/components/tests/test_complete_condenser.py:53]: Should there be another test file for `compressor`?

**Review by adam-a-a — COMMENTED** [review]
I had some questions about the implementation of unit models. @bknueven seems to have some important comments on initialization routines that would need to be addressed.

File: watertap/unit_models/mvc/tests/test_mvc.py (hunk 3 above)
Comment (carson-tucker) [reply watertap/unit_models/mvc/tests/test_mvc.py:55]: This compressor model is made to be used with the water property package which does not contain specific entropy as a property. The isentropic temperature of the outlet is found assuming ideal gas.

File: watertap/unit_models/mvc/tests/test_mvc.py (hunk 4 above)
Comment (carson-tucker) [reply watertap/unit_models/mvc/tests/test_mvc.py:58]: This condenser model was written to be used with the water property package and requires that the outlet is fully condensed.

File: watertap/unit_models/mvc/tests/test_mvc.py (hunk 5 above)
Comment (carson-tucker) [reply watertap/unit_models/mvc/tests/test_mvc.py:66]: The previous evaporator model contained the evaporator and condenser as a single unit model, but this was challenging to initialize. The unit models are also separate to allow for multi-stage MVC flowsheets.

File: watertap/unit_models/mvc/components/complete_condenser.py (hunk 1 above)
Comment (carson-tucker) [reply watertap/unit_models/mvc/components/complete_condenser.py]: I changed the logic and the name.

File: watertap/unit_models/mvc/components/evaporator.py (hunk 2 above)
Comment (carson-tucker) [reply watertap/unit_models/mvc/components/evaporator.py]: I changed the logic and the name.

**Review by bknueven — APPROVED** [review]

File: watertap/unit_models/mvc/components/complete_condenser.py [hunk 7]
```diff
@@ -253,6 +256,24 @@ def initialize(
...
+        # check if guess is needed for the heat based on degrees of freedom
+        has_guessed_heat = False
+        if degrees_of_freedom(blk) > 1:
+            raise RuntimeError(
+                "The model has {} degrees of freedom rather than 0 or 1 (with a guessed heat) for initialization."
+                " This error suggests that an outlet condition has not been fixed"
+                " for initialization.".format(degrees_of_freedom(blk))
+            )
+        elif degrees_of_freedom(blk) == 1:
+            if heat != None:
```
Comment (bknueven) [inline watertap/unit_models/mvc/components/complete_condenser.py]:
```suggestion
            if heat is not None:
```

File: watertap/unit_models/mvc/components/evaporator.py [hunk 8]
```diff
@@ -430,70 +429,60 @@ def initialize(
...
-        blk.condenser.initialize(state_args=state_args_condenser)
-        # assert False
-        # flags_condenser_cv = blk.condenser.initialize(state_args=state_args_condenser,hold_state=True)
-        init_log.info_high("Initialization Step 3 Complete.")
-        # ---------------------------------------------------------------------
-        # Deactivate heat transfer balance
-        # blk.eq_heat_balance.deactivate()
+        # incorporate guessed temperature differences
+        has_guessed_delta_temperature_in = False
+        if delta_temperature_in != None:
```
Comment (bknueven) [inline watertap/unit_models/mvc/components/evaporator.py]:
```suggestion
        if delta_temperature_in is not None:
```

File: watertap/unit_models/mvc/components/evaporator.py [hunk 9]
```diff
@@ -430,70 +429,60 @@ def initialize(
...
+                raise RuntimeError(
+                    "A guess was provided for the delta_temperature_in variable in the "
+                    "initialization, but it is already fixed. Either do not "
+                    "provide a guess for or unfix delta_temperature_in"
+                )
+            blk.delta_temperature_in.fix(delta_temperature_in)
+            has_guessed_delta_temperature_in = True
+
+        has_guessed_delta_temperature_out = False
+        if delta_temperature_out != None:
```
Comment (bknueven) [inline watertap/unit_models/mvc/components/evaporator.py]:
```suggestion
        if delta_temperature_out is not None:
```

**Review by bknueven — COMMENTED** [review]
The initialization logic is clear and improved, thanks.

I have a few nits on comparisons against `None`; see http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html

**Review by TimBartholomew — APPROVED** [review]
LGTM

**Pull request merged**